src/function/numeric/solveODE.js

```diff
@@ -0,0 +1,254 @@
+     * @return {Object} Return an object with t and y values as arrays
+     */
+
+  function _rk (butcherTableau) {
+    // generates an adaptive runge kutta method from it's butcher tableau
+
+    return function (f, T, y0, options) {
+      // adaptive runge kutta methods
+
+      const t0 = T[0] // initial time
+      const tf = T[1] // final time
+      const steps = 1 // divide the time in this number of steps
+      const tol = options.tol ? options.tol : 1e-4 // define a tolerance (must be an option)
+      const maxStep = options.maxStep
+      const minStep = options.minStep
+      const minDelta = options.minDelta ? options.minDelta : 0.2
+      const maxDelta = options.maxDelta ? options.maxDelta : 5
+      const maxIter = options.maxIter ? options.maxIter : 10_000 // stop inifite evaluation if something goes wrong
+      const [a, c, b, bp] = [butcherTableau.a, butcherTableau.c, butcherTableau.b, butcherTableau.bp]
+
+      let h = options.firstStep
+        ? options.firstStep
+        : divide(subtract(tf, t0), steps) // define the first step size
+      const t = [t0] // start the time array
+      const y = [y0] // start the solution array
+
+      const dletaB = subtract(b, bp) // b - bp
+
+      let n = 0
+      let iter = 0
+      // iterate unitil it reaches either the final time or maximum iterations
+      while (ongoing(t[n], tf, h) && iter < maxIter) {
+        const k = []
+
+        // trim the time step so that it doesn't overshoot
+        h = trimStep(t[n], tf, h)
+
+        // calculate the first value of k
+        k.push(f(t[n], y[n]))
+
+        // calculate the rest of the values of k
+        for (let i = 1; i < c.length; ++i) {
+          k.push(
+            f(
+              add(t[n], multiply(c[i], h)),
+              add(y[n], multiply(h, a[i], k))
+            )
+          )
+        }
+
+        // estimate the error by comparing solutions of different orders
+        const TE = max(
+          abs(
+            map(multiply(dletaB, k), (X) =>
+              isUnit(X) ? X.value : X
+            )
+          )
+        )
+
+        if (TE < tol && tol / TE > 1 / 4) {
+          // if within tol then push everything
+          t.push(add(t[n], h))
+          y.push(add(y[n], multiply(h, b, k)))
+          n++
+        }
+
+        // estimate the delta value that will affect the step size
+        const delta = 0.84 * (tol / TE) ** (1 / 5)
+
+        if (delta < minDelta) {
+          h = multiply(h, minDelta)
+        } else if (delta > maxDelta) {
+          h = multiply(h, maxDelta)
+        } else {
+          h = multiply(h, delta)
+        }
+
+        if (maxStep && larger(abs(h), abs(maxStep))) {
+          h = isPositive(h) ? abs(maxStep) : multiply(-1, abs(maxStep))
+        } else if (minStep && smaller(abs(h), abs(minStep))) {
+          h = isPositive(h) ? abs(minStep) : multiply(-1, abs(minStep))
+        }
+        iter++
+      }
+      return { t, y }
```

Comment:
Thinking aloud here: should the function throw an error when it didn't converge and reached `maxIter`?

Reply:
Yes, currently it might show an incomplete solution, but maybe it should be an error or an empty solution with an error. I will search in other functions for error references.

Reply:
also implemented